lib/nx_signal.ex

```diff
@@ -165,44 +201,77 @@ defmodule NxSignal do
           [0, 0]
         ]
       >
+
+      iex> t = Nx.iota({7});
+      iex> NxSignal.as_windowed(t, window_size: 6, padding: :reflect, stride: 1)
+      #Nx.Tensor<
+        s64[7][6]
+        [
+          [1, 2, 1, 0, 1, 2],
+          [2, 1, 0, 1, 2, 3],
+          [1, 0, 1, 2, 3, 4],
+          [0, 1, 2, 3, 4, 5],
+          [1, 2, 3, 4, 5, 6],
+          [2, 3, 4, 5, 6, 5],
+          [3, 4, 5, 6, 5, 4]
+        ]
+      >
   """
-  defn as_windowed(tensor, opts \\ []) do
-    # current implementation only supports windowing 1D tensors
-    {window_size, stride, padding, output_shape} =
-      transform({Nx.shape(tensor), opts}, fn {shape, opts} ->
-        opts = Keyword.validate!(opts, [:window_size, padding: :valid, stride: 1])
-        window_size = opts[:window_size]
-        window_dimensions = {window_size}
-
-        padding = opts[:padding]
-
-        [stride] =
-          strides =
-          case opts[:stride] do
-            stride when Elixir.Kernel.is_list(stride) ->
-              stride
-
-            stride
-            when Elixir.Kernel.and(
-                   Elixir.Kernel.is_integer(stride),
-                   Elixir.Kernel.>=(stride, 1)
-                 ) ->
-              [stride]
-
-            stride ->
-              raise ArgumentError,
-                    "expected an integer >= 1 or a list of integers, got: #{inspect(stride)}"
-          end
+  deftransform as_windowed(tensor, opts \\ []) do
+    if opts[:padding] == :reflect do
+      as_windowed_reflect_padding(tensor, opts)
+    else
+      as_windowed_non_reflect_padding(tensor, opts)
+    end
+  end
+
+  deftransformp as_windowed_parse_opts(shape, opts, :reflect) do
+    window_size = opts[:window_size]
 
-        dilations = List.duplicate(1, Nx.rank(tensor))
+    as_windowed_parse_opts(
+      shape,
+      Keyword.put(opts, :padding, [{div(window_size, 2), div(window_size, 2) - 1}])
+    )
+  end
 
-        {pooled_shape, padding_config} =
-          Nx.Shape.pool(shape, window_dimensions, strides, padding, dilations)
+  deftransformp as_windowed_parse_opts(shape, opts) do
+    opts = Keyword.validate!(opts, [:window_size, padding: :valid, stride: 1])
+    window_size = opts[:window_size]
+    window_dimensions = {window_size}
 
-        output_shape = {Tuple.product(pooled_shape), window_size}
+    padding = opts[:padding]
 
-        {window_size, stride, Enum.map(padding_config, fn {x, y} -> {x, y, 0} end), output_shape}
-      end)
+    [stride] =
+      strides =
+      case opts[:stride] do
+        stride when Elixir.Kernel.is_list(stride) ->
+          stride
+
+        stride
+        when Elixir.Kernel.and(
+               Elixir.Kernel.is_integer(stride),
+               Elixir.Kernel.>=(stride, 1)
+             ) ->
+          [stride]
```

Comment:
Why are you using `Elixir.Kernel` here? :)

Reply:
Probably this was a `transform` before with inline `fn`

Reply:
No need for this now